sgkit/stats/regenie.py

```diff
@@ -0,0 +1,778 @@
+    (array([0, 2]), array([2, 1]))
+    >>> index_array_blocks([0, 0, 1, 1, 1], 2)
+    (array([0, 2, 4]), array([2, 2, 1]))
+
+    Returns
+    -------
+    index : ndarray
+        Array of indexes for each block start
+    sizes : ndarray
+        Size of block such that `x[index[0]:(index[0] + sizes[0])]` contains
+        every element in block 0
+
+    Raises
+    ------
+    ValueError
+        If `x` is not 1D.
+    ValueError
+        If `size` is <= 0.
+    ValueError
+        If `x` does not contain integers.
+    ValueError
+        If `x` is not monotonic increasing.
+    """
+    x = np.asarray(x)
+    if x.size == 0:
+        return np.empty(0, dtype=int), np.empty(0, dtype=int)
+    if x.ndim != 1:
+        raise ValueError(f"Array shape {x.shape} is not 1D")
+    if size <= 0:
+        raise ValueError(f"Block size {size} must be > 0")
+    if not np.issubdtype(x.dtype, np.integer):
+        raise ValueError("Array to partition must contain integers")
+    if np.any(np.diff(x) < 0):
+        raise ValueError("Array to partition must be monotonic increasing")
+    breaks = np.argwhere(np.diff(x, prepend=x[0]))[:, 0]
+    breaks = np.concatenate(([0], breaks, [x.size]))
+    index = np.concatenate(
+        [np.arange(breaks[i], breaks[i + 1], size) for i in range(breaks.size - 1)]
+    )
+    sizes = np.diff(index, append=x.size)
+    assert index.size == sizes.size
+    return index, sizes
+
+
+def index_block_sizes(
+    sizes: Union[ArrayLike, Sequence[int]]
+) -> Tuple[ndarray, ndarray]:
+    """Generate indexes for blocks of specific sizes.
+
+    Parameters
+    ----------
+    sizes : Union[ArrayLike, Sequence[int]]
+        Block sizes to generate indexes for.
+
+    Examples
+    --------
+    >>> from sgkit.stats.regenie import index_block_sizes
+    >>> index_block_sizes([3, 4, 5])
+    (array([0, 3, 7]), array([3, 4, 5]))
+
+    Returns
+    -------
+    index : ndarray
+        Array of indexes for each block start.
+    sizes : ndarray
+        Size of block such that `x[index[0]:(index[0] + sizes[0])]` contains
+        every element in block 0.
+
+    Raises
+    ------
+    ValueError
+        If any value in `sizes` is <= 0.
+    ValueError
+        If `sizes` does not contain integers.
+    """
+    sizes = np.asarray(sizes)
+    if np.any(sizes <= 0):
+        raise ValueError("All block sizes must be >= 0")
+    if not np.issubdtype(sizes.dtype, np.integer):
+        raise ValueError("Block sizes must be integers")
+    chunks = np.concatenate([np.array([0]), sizes])
+    index = np.cumsum(chunks)[:-1]
+    assert index.size == sizes.size
+    return index, sizes
+
+
+def ridge_regression(
+    XtX: ArrayLike,
+    XtY: ArrayLike,
+    alphas: Union[ArrayLike, Sequence[float]],
+    n_zero_reg: Optional[int] = None,
+    dtype: Any = None,
+) -> ArrayLike:
+    """Multi-outcome, multi-parameter ridge regression from CV intermediates."""
+    assert XtX.shape[0] == XtX.shape[1] == XtY.shape[0]
```

Comment:
Should this raise `ValueError`?

Reply:
Maybe?  It definitely shouldn't be a publicly facing method and I'm not really avoiding hidden methods for a good reason other than that I don't like making it harder to find for people that want to investigate the implementation.  That raises a good question though in what we should expose within modules even if they aren't exposed in the top-level API.  Either way it should have at least have an informative error so I'll change that, though my general thinking is that assertions are fine when validating some invariant that can't be violated based on user input.  This would be one of those cases, but I can see how it's not clear given that it seems like the method might not be used only internally.

Reply:
I changed it though it's not in general a practice I was going to stick with: https://github.com/pystatgen/sgkit/pull/66/commits/64c1da651f5cd2596f28461639981809f80519ad#diff-74c0b482520e25ae56fbd4e900dede59R138

Perhaps we make some rule that all argument validation works without assertions unless it's a hidden method?  That would be an argument for me hiding several of these currently unhidden functions.  I don't particularly care either way.